MPT review

CeciliaZ030 opened the pull request:
[MPT review #39](https://github.com/taikoxyz/zkevm-circuits/issues/39)

CeciliaZ030 committed:
constrain ext_is_key_part_odd with fixed lookup

CeciliaZ030 committed:
add lookup for RlpGadgets

CeciliaZ030 committed:
Revert "constrain ext_is_key_part_odd with fixed lookup"

This reverts commit 735d2bf4e72f1f909853538dd164b8822fce536d.

CeciliaZ030 committed:
is_key_part_odd with lookups done

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -934,11 +1005,14 @@ mod tests {
 
     #[test]
     fn test_mpt() {
+        let only_run = var("ONLY_RUN")
+            .and_then(|idx| idx.parse::<usize>().map_err(|e|VarError::NotPresent)
+        ).ok();
+        println!("{:?}", only_run);
```
Honestly for some reason I hadn't though of doing this myself, but this option makes a lot of sense!

CeciliaZ030 replied:
Missing `is_long` is because I thought we can represent the status of rlp values with two bytes:
```
 is_short   is_very_long
     true          false        short
     false         false        long
     false.        true        very_long
```
Isn't it better to same some Fixed Column? :)

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -870,6 +874,73 @@ impl<F: Field> MPTConfig<F> {
                     offset += 1;
                 }
 
+                // Rlp prefixes table [rlp_tag, byte, is_string, is_short, is_verylong]
+                for ind in 0..=127 {
```
Could you do this a bit more like in the witness generation with a match statement for example. Will be quite a bit shorter, less duplicated code (like setting the tag to the same value for each case) and that also uses the RLP constants like `RLP_SHORT` etc... so fewer magic values.

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -870,6 +874,73 @@ impl<F: Field> MPTConfig<F> {
+
+                // KEY_PREFIX_ODD: u8 = 0b0001_0000;
+                // first byte of ext key >> 4 == 1
+                for ind in 0..256 {
+                    assignf!(region, (self.fixed_table[0], offset) => FixedTableTag::ExtOddKey.scalar())?;
+                    assignf!(region, (self.fixed_table[1], offset) => ind.scalar())?;
+                    if 16 <= ind && ind < 32 {
+                        assignf!(region, (self.fixed_table[2], offset) => true.scalar())?;
+                    } else {
+                        assignf!(region, (self.fixed_table[2], offset) => false.scalar())?;
+                    }
+                    offset += 1;
+                }
```
This works as expected, but I checked the compact encoding spec again and it seems like to enforce the correct encoding we can limit the possible values so the circuit has to do less of that:
```
// Even - only the nibbles 0 0 are valid
assignf!(region, (self.fixed_table[0], offset) => FixedTableTag::ExtOddKey.scalar())?;
assignf!(region, (self.fixed_table[1], offset) => 0.scalar())?;
assignf!(region, (self.fixed_table[2], offset) => false.scalar())?;
offset += 1;
// Odd - First nibble is 1, the second nibble can be any value
for idx in 0..16 {
    assignf!(region, (self.fixed_table[0], offset) => FixedTableTag::ExtOddKey.scalar())?;
    assignf!(region, (self.fixed_table[1], offset) => ((0b1_0000) + idx).scalar())?;
    assignf!(region, (self.fixed_table[2], offset) => true.scalar())?;
    offset += 1;
}

CeciliaZ030 replied:
Yea you're right, I saw this [https://github.com/privacy-scaling-explorations/zkevm-specs/blob/21640e4834a5eb36fc2fb4f4108520d1c4f4f655/specs/mpt/extension-node.md?plain=1#L237](url).
Will make the change 👌

Brechtpd commented on zkevm-circuits/src/mpt_circuit/rlp_gadgets.rs:
```diff
@@ -37,12 +40,26 @@ pub(crate) struct RLPListWitness {
+            require!(vec![
+                FixedTableTag::RLP.expr(), 
+                bytes[0].clone(), 
+                is_string.expr(), 
+                is_short.expr(), 
+                is_very_long.expr()] => @"fixed"
```
`is_long` seems to be missing? I think that's actually smart to only do 2 out of 3 in the lookup table because it saves a cell and you can deduce one of the three values e.g. `is_long` is `1 - is_short - is_very_long`, but I don't see that being done so not sure what's up? :)

Brechtpd reviewed (commented):
Generally looks good! Just some suggestions and a missing `is_long` in the lookup table that needs some explanation/implementation to make it work.

CeciliaZ030 committed:
added q_row and q_node to constraint state

CeciliaZ030 committed:
constraint nodes

Brechtpd commented on zkevm-circuits/src/mpt_circuit/param.rs:
```diff
@@ -143,6 +143,12 @@ pub const RLP_HASH_VALUE: u8 = 128 + 32; //  0x80
 pub const KEY_LEN_IN_NIBBLES: usize = KEY_LEN * 2;
 
+// MPT row constraints
+pub const START_NODE_ROWS: usize = 2;
+pub const BRANCH_NODE_ROWS: usize = 21;
+pub const ACCOUNT_NODE_ROWS: usize = 12;
+pub const LEAF_NODE_ROWS: usize = 6;
```
These are also directly available (and automatically always have the correct value) by using the enums: `StorageRowType::Count`, `AccountRowType::Count`, ...

CeciliaZ030 replied:
Ahh right, forgot about them.

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -205,18 +237,42 @@ impl<F: Field> MPTConfig<F> {
                         },
                         a!(is_branch) => {
+                            require!((FixedTableTag::BranchNode.expr(), a!(q_row)) => @"fixed");
+                            let diff = a!(q_node)- a!(q_node_prev);
+                            // Start -> Branch || Branch -> Branch
+                            require!(
+                                (diff.clone() - START_NODE_ROWS.expr()) * (diff.clone() - BRANCH_NODE_ROWS.expr()) 
+                                => 0.expr());
```
Doing it like this is a bit tricky. I think it's better to just make sure for each node that `a!(is_start), a!(is_branch), a!(is_account) or a!(is_storage)` are set to `1` after `Count` rows. So for example for a branch node, for all following rows up to `ExtensionBranchRowType::Count`, `a!(is_start) + a!(is_branch) + a!(is_account) + a!(is_storage) => 0`, on the `ExtensionBranchRowType::Count`th row however, `a!(is_start) + a!(is_branch) + a!(is_account) + a!(is_storage) => 1`.

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -205,18 +237,42 @@ impl<F: Field> MPTConfig<F> {
                     ifx! {not!(f!(q_not_first)) => {
                         require!(a!(is_start) => true);
                     }};
+                    cb.base.require_zero(
+                        "only one state is valid at a time",
+                        a!(is_start) * a!(is_branch) * a!(is_account) * a!(is_storage)
+                    );
```
I think you mean `a!(is_start) + a!(is_branch) + a!(is_account) + a!(is_storage) => 1` because what you're checking here will be valid as long as only 1 is non-zero.

Also, the good news is that `matchx!` will kind of already do this check (as well as check that those values are boolean)! I say kind of because the `matchx!` does have a `_` so they are are also all allowed to be `0`.

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -205,18 +237,42 @@ impl<F: Field> MPTConfig<F> {
                     matchx! {
                         a!(is_start) => {
+                            require!(a!(q_node) => 0.expr());
+                            require!((FixedTableTag::StartNode.expr(), a!(q_row)) => @"fixed");
```
Why does this use a lookup?

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -735,17 +798,22 @@ impl<F: Field> MPTConfig<F> {
                         for (byte, &column) in bytes.iter().zip(self.main.bytes.iter()) {
                             assign!(region, (column, offset + idx) => byte.scalar())?;
                         }
+                        assign!(region, (self.q_row, offset + idx) => idx.scalar())?;
+                        assign!(region, (self.q_node, offset + idx) => offset.scalar())?;
+                        assign!(region, (self.q_node_prev, offset + idx) => offset_prev.scalar())?;
```
Keeping the total row idx in `q_node` + `q_row` is possible, but I don't see that much benefit over it compared to always starting from `0` or `RowType::Count` for each row? I guess it's quite similar, but seems less state to keep track of that way.

CeciliaZ030 replied:
Yeah, all my efforts was for constraining `Start -> Branch* -> Account -> Branch* -> Storage` with their designated rows offsets. But since you explained that the circuits have these constraint already with `is_below_account`, I guess these are not necessary...

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -870,6 +939,90 @@ impl<F: Field> MPTConfig<F> {
+                    offset += 1;
+                }
+
+                for ind in 0..2 {
+                    assignf!(region, (self.fixed_table[0], offset) => FixedTableTag::StartNode.scalar())?;
+                    assignf!(region, (self.fixed_table[1], offset) => ind.scalar())?;
+                }
+                for ind in 0..21 {
+                    assignf!(region, (self.fixed_table[0], offset) => FixedTableTag::BranchNode.scalar())?;
+                    assignf!(region, (self.fixed_table[1], offset) => ind.scalar())?;
+                }
+                for ind in 0..12 {
+                    assignf!(region, (self.fixed_table[0], offset) => FixedTableTag::AccountNode.scalar())?;
+                    assignf!(region, (self.fixed_table[1], offset) => ind.scalar())?;
+                }
+                for ind in 0..6 {
+                    assignf!(region, (self.fixed_table[0], offset) => FixedTableTag::StorageNode.scalar())?;
+                    assignf!(region, (self.fixed_table[1], offset) => ind.scalar())?;
+                }
```
I'm not sure I actually fully understand the lookup use for this (seems easily replaceable by a hard-coded constant in the circuit for the `Count`) but this doesn't update `offset` so this will overwrite the same row over and over again.

CeciliaZ030 replied:
My bad, the `offset` is a mistake. In general since I took the `q_row` approach, I was trying to constrain `q_row` dynamically so that in each state it's in a designated range, but I wanted to avoid using a lot of columns to track differences in offsets. But I'll remove these since you clarify with the table that has `is_start`, `is_branch`, `is_account`, and `is_storage`

Brechtpd reviewed (commented):
I think you got the correct idea for the state machine selectors, but seems like the implementation is more complex than needed.

A table so it's easier to talk about:

| row | is_start | is_branch | is_account | is_storage |
|------|-----------|---------------|----------------|---------------|
|  0     |      1     |       0         |        0         |        0        |
|  1     |      0     |       0         |        0         |        0        |
|  2     |      0     |       1         |        0         |        0        |
|  3     |      0     |       0         |        0         |        0        |
| 4-23 |      0     |       0         |        0         |        0        |
|  24   |       0    |       0         |         1        |        0        |
| ...   |      ...   |      ...         |         ...        |       ....        |

So this is the current state of the circuit. The selectors for each node are enabled on each row correctly by the witness, but the circuit doesn't have any checks that enforces this.

So I think the only thing that really needs to be checked is that for e.g. the first row with a start node (which has StartRowType::Count rows) has zero's for all the state selectors on the next row, but on the 2nd row there is exactly one selector enabled again (can be any type, doesn't matter).

And because on each row we do check that the selectors are boolean, we can pretty much just brute-force this check
- For all state selectors in the following rows up to `RowType::Count`, basically just sum all those selectors togethers and force the sum to be `0`.
- On the ``RowType::Count`th, sum the selectors just on that row and check that it's `1`.

Easy solution, and pretty efficient.

The EVM circuit instead of doing the brute force method does do something similar you're trying to do here with counting down, so e.g.

| row | is_start | is_branch | is_account | is_storage |  num_data_rows_left |
|------|-----------|---------------|----------------|---------------|----------------------------------|
|  0     |      1     |       0         |        0         |        0        |                0
|  1     |      0     |       0         |        0         |        0        |  StartRowType::Count - 1
|  2     |      0     |       1         |        0         |        0        |                 0
|  3     |      0     |       0         |        0         |        0        |  BranchRowType::Count - 1

Where the number of data only rows are written into `num_data_rows_left` and then there's 2 cases:
- `num_data_rows_left == 0` -> A state selector needs to be enabled on the current row
- `num_data_rows_left != 0` -> All state selectors need to be disabled on the current row, and `num_data_rows_left.cur() - 1 => num_data_rows_left.next()`.

CeciliaZ030 committed:
added constraint for Count-th rows and non-Count-th rows && removed lookup for state machine

CeciliaZ030 committed:
Chang nibble rules 0b1_0000

CeciliaZ030 commented:
I removed the unnecessary lookups as we discussed 😊 and added checks according to your [table](https://github.com/Brechtpd/zkevm-circuits/pull/9#pullrequestreview-1346225167).
One thing is that I decided to leave `Start -> Branch* -> Account -> Branch* -> Storage` with `q_row` and `q_node` constraints because the current `is_below_branch` only make sure no account is beneath account. I feel like is better to have a set of holistic state machine constraints.

CeciliaZ030 committed:
Chang nibble rules 0b1_0000

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -205,21 +238,43 @@ impl<F: Field> MPTConfig<F> {
+                    // When q_row > 0, we're in the middle of some node rows, all flags needs to be 0 
+                    // Otherwise q_row == 0, we start at a new node and corresponding flag needs to be 1
+                    // taken care by  _ => require!(true => false)
+                    ifx! {a!(q_row)  => {
```
Are there currently constraints that update `q_row` on each row? `q_row` also counts up, but how does the circuit knows when to allow the next state selector to be set in this case? (compared to counting down, where the state can set the number of columns until the next state needs to be enabled when the counter is back to `0`)

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -205,21 +238,43 @@ impl<F: Field> MPTConfig<F> {
                         a!(is_branch) => {
+                            // q_node[cur] - count == q_node[cur-count]
+                            // Start -> Branch || Branch -> Branch
+                            let diff1 = a!(q_node) - (ExtensionBranchRowType::Count as i32).expr() - a!(q_node, -(ExtensionBranchRowType::Count as i32));
```
I'm a bit surprised this rotation works to be honest. The difficulty with rotation to previous rows is that we have to check that the rotation on the current row doesn't get to some row < 0. This actually wraps around the circuit (so a rotation of -1 on the first row accesses the data on the last row), but the additional difficulty is that on the last rows of the circuit there's random data stored to enable zk. A branch is almost always used after the initial start which uses 2 rows, and `ExtensionBranchRowType::Count` is a pretty large value, so this seems to read from this garbage data (halo2 calls these poisoned cells, and the rows containing the random data unusable rows).

So most of the time you actually want to avoid using previous values because they are trickier, it's easier (and basically equivalent) to only look to the next rows instead.

CeciliaZ030 replied:
> The difficulty with rotation to previous rows is that we have to check that the rotation on the current row doesn't get to some row < 0. This actually wraps around the circuit (so a rotation of -1 on the first row accesses the data on the last row), but the additional difficulty is that on the last rows of the circuit there's random data stored to enable zk.

For the specific arrangement dealing with rotation, does halo2 have docs for it?

Brechtpd replied:
Not really any good ones that I know of, they do touch on it in a couple of places, for example:
- https://zcash.github.io/halo2/design/proving-system/lookup.html?highlight=blinding#zero-knowledge-adjustment
- https://github.com/privacy-scaling-explorations/halo2/blob/fcdd5b910d92ab278d679cf8aa176c885455aa6c/halo2_proofs/src/plonk/circuit.rs#L2011
- https://github.com/privacy-scaling-explorations/halo2/blob/fcdd5b910d92ab278d679cf8aa176c885455aa6c/halo2_proofs/src/dev.rs#L636 (this it where the cell values are set as "poisoned" in the MockProver so it will return an error).

Brechtpd commented:
>One thing is that I decided to leave `Start -> Branch* -> Account -> Branch* -> Storage` with `q_row` and `q_node` constraints because the current `is_below_branch` only make sure no account is beneath account. I feel like is better to have a set of holistic state machine constraints.

We'll see. ;) I also think that generally you do indeed want to add these, but in this case it's a bit tricky to get right. Like `Account -> Branch -> Storage` is not the only option I believe, it's also possible to go directly `Account -> Storage` when there's only a single storage leaf. Same with the accounts I think.

These are all the valid transitions I think:
```
- Start -> Branch
- Start -> Account
- Start -> Start (for padding purposes)

- Branch -> Branch
- Branch -> Account
- Branch -> Storage
- Branch -> Start (doesn't make much sense, but because Start resets everything we can allow it)
- 
- Account -> Storage
- Account -> Start
- Account -> Branch

- Storage -> Start
- Storage -> X (could be made valid if we reset the necessary state as well)
```

So most transitions are actually allowed by the state machine itself, but we need checks elsewhere where we have more context what's going on to actually check if the transition make sense.

CeciliaZ030 committed:
got rid of state machine and constrain q_row

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -192,35 +229,70 @@ impl<F: Field> MPTConfig<F> {
+                let is_q_row_zero = IsZeroConfig {
+                    value_inv: ctx.q_row_inv.clone(),
+                    is_zero_expression: 1.expr() - a!(q_row) * a!(q_row_inv),
+                };
```
You can use the `IsZeroGadget` which integrates directly with the ConstraintBuilder: https://github.com/Brechtpd/zkevm-circuits/blob/67d68e8e21524e28e88a5baff32cacc4c5a524c9/zkevm-circuits/src/circuit_tools/gadgets.rs#L20

CeciliaZ030 replied:
Will do :)

CeciliaZ030 committed:
query_byte() with lookup

CeciliaZ030 committed:
use isZeroGadget, add mpt_test2 for VS code debugger

CeciliaZ030 committed:
broken CachedRegion

CeciliaZ030 committed:
use CachedRegion for all states

CeciliaZ030 committed:
add split_expression

CeciliaZ030 committed:
added table

CeciliaZ030 committed:
fixed byte lookup with a separate column
added  APIs for MPTConstraintBuilder similar to evm_circuit

CeciliaZ030 committed:
load_byte_table

CeciliaZ030 committed:
Smarter cell manager with context

CeciliaZ030 committed:
use tree-like structure to store context
put in ifx macros but not working yet

Brechtpd commented on zkevm-circuits/src/circuit_tools/cell_manager.rs:
```diff
@@ -148,22 +188,243 @@ pub struct CellColumn<F> {
+    cells: Vec<Cell<F>>,
+    columns: Vec<CellColumn<F>>,
+    // branch ctxs
+    branch_ctxs: HashMap<String, CmContext<F>>,
```
Isn't some kind of stack more convenient? So you can just push and pop contexts as needed without having to worry about parents and stuff.

Brechtpd commented on zkevm-circuits/src/mpt_circuit.rs:
```diff
@@ -180,9 +181,14 @@ impl<F: Field> MPTConfig<F> {
         memory.allocate(meta, main_memory());
 
-        let ctx = MPTContext {
+        let mut cb = MPTConstraintBuilder::new(33 + 10, None, power_of_randomness.clone());
```
```suggestion
        let mut cb = MPTConstraintBuilder::new(9, None, power_of_randomness.clone());
```

This should be a good test so see if the expression splitting (and CachedRegion) is working correctly! (you may need to allocate some extra columns to the cell manager because this may need quite a few extra cells to store intermediate results).

Brechtpd commented on zkevm-circuits/src/mpt_circuit/helpers.rs:
```diff
@@ -30,8 +32,17 @@ use super::{
+    table::Lookup,
 };
 
+// Max degree allowed in all expressions passing through the ConstraintBuilder.
+// It aims to cap `extended_k` to 2, which allows constraint degree to 2^2+1,
+// but each ExecutionGadget has implicit selector degree 3, so here it only
+// allows 2^2+1-3 = 2.
+const MAX_DEGREE: usize = 5;
+const IMPLICIT_DEGREE: usize = 3;
```
This is EVM circuit specific so to generalize things don't use these anymore. Instead of `MAX_DEGREE` use `max_degree` as stored in the ConstraintBuilder. `IMPLICIT_DEGREE` you can drop completely, we shouldn't need it anymore.

Brechtpd reviewed (commented):
Went over things pretty quickly but generally looking good!

Not sure why you kept the non-`CachedRegion` `assign` functions for now, but you can remove if you're done converting everything to the `CachedRegion` ones.

CeciliaZ030 commented:
> Not sure why you kept the non-CachedRegion assign functions for now, but you can remove if you're done converting everything to the CachedRegion ones.

I tried to code in a way that if I break something I can use old things :) And I changed some of the names (like _CellManager  with the dash) so I can change all occurance easily.

CeciliaZ030 committed:
changed macros to enable CM context for _ifx and _matchx
